### PR TITLE
Expose HostMetrics in OkHttpClients

### PR DESCRIPTION
**Pull request by ellisjoe**

### CODE REVIEW

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/metrics/HostMetrics.java [hunk 1]
```diff
@@ -39,16 +42,18 @@
...
-        successful = registry.meter(name(serviceName, hostname, "successful"));
-        redirection = registry.meter(name(serviceName, hostname, "redirection"));
-        clientError = registry.meter(name(serviceName, hostname, "client-error"));
-        serverError = registry.meter(name(serviceName, hostname, "server-error"));
-        other = registry.meter(name(serviceName, hostname, "other"));
+    HostMetrics(TaggedMetricRegistry registry, String serviceName, String hostname) {
+        this.serviceName = serviceName;
+        this.hostname = hostname;
+        this.informational = registry.meter(name("informational"));
+        this.successful = registry.meter(name("successful"));
```
Comment (ellisjoe) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/metrics/HostMetrics.java]: should these be timers?

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/metrics/HostMetricsRegistry.java [hunk 2]
```diff
@@ -20,34 +20,48 @@
...
+    public void record(String serviceName, String hostname, int statusCode) {
         try {
-            hostMetrics.getUnchecked(hostname).record(statusCode);
+            hostMetrics.getUnchecked(ImmutableServiceAndHost.of(serviceName, hostname)).record(statusCode);
         } catch (Exception e) {
             log.warn("Unable to record metrics for host", UnsafeArg.of("hostname", hostname));
         }
     }
+
+    public Collection<HostMetrics> getMetrics() {
```
Comment (ellisjoe) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/metrics/HostMetricsRegistry.java]: possibly worth exposing the map here. means we can't record things finer grained that the service/host level.

**Comment by uschi2000** [comment]
rebase, please

**Comment by ellisjoe** [comment]
@uschi2000 rebased

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/OkHttpClients.java [hunk 3]
```diff
@@ -81,6 +91,13 @@ public static OkHttpClient create(ClientConfiguration config, String userAgent,
         return create(config, UserAgents.tryParse(userAgent), serviceClass);
     }
 
+    /**
+     * Return the per service and host metrics for all clients created by {@link OkHttpClients}.
+     */
+    public static Collection<HostMetrics> hostMetrics() {
```
Comment (ellisjoe) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/OkHttpClients.java:95]: main api concerns are this method and the changes to `HostMetrics`

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/OkHttpClients.java [hunk 4]
```diff
@@ -60,6 +64,12 @@
     private static final ScheduledExecutorService scheduledExecutorService = Tracers.wrap(
             new ScheduledThreadPoolExecutor(5, Util.threadFactory("http-remoting/OkHttp Scheduler", false)));
 
+    /**
+     * The per service and host metrics recorded for each HTTP call.
+     */
+    private static final HostMetricsRegistry hostMetrics =
```
Comment (uschi2000) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/OkHttpClients.java:68]: would be awesome if could inject this. possible?

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/metrics/HostMetricsRegistry.java [hunk 5]
```diff
@@ -20,34 +20,48 @@
...
-                    public HostMetrics load(String hostname) throws Exception {
-                        return new HostMetrics(registry, serviceName, hostname);
+                    public HostMetrics load(ServiceAndHost serviceAndHost) throws Exception {
+                        return new HostMetrics(registry, serviceAndHost.serviceName(), serviceAndHost.hostname());
                     }
                 });
     }
 
-    public void record(String hostname, int statusCode) {
+    public void record(String serviceName, String hostname, int statusCode) {
```
Comment (uschi2000) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/metrics/HostMetricsRegistry.java]: little dangerous to change this api

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/OkHttpClients.java (hunk 4 above)
Comment (ellisjoe) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/OkHttpClients.java:68]: if we wanna start passing around instances of `OkHttpClients` and deprecate the static `create` method. I'm not convinced being able to pass one of these in is particularly useful?

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/metrics/HostMetricsRegistry.java (hunk 5 above)
Comment (ellisjoe) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/metrics/HostMetricsRegistry.java]: I'd say it's small enough. this has been in 1 release? also this is really internal api

Comment (markelliot) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/metrics/HostMetricsRegistry.java]: possible to make package private?

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/InstrumentedInterceptor.java [hunk 6]
```diff
@@ -29,22 +26,24 @@
 public final class InstrumentedInterceptor implements Interceptor {
```
Comment (markelliot) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/InstrumentedInterceptor.java]: this doesn't need to be public yeah?

Comment (ellisjoe) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/InstrumentedInterceptor.java]: done

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/DefaultHostMetrics.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+import com.codahale.metrics.Timer;
+import com.palantir.tritium.metrics.registry.MetricName;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Records per-target-host HTTP response code metrics in a {@link TaggedMetricRegistry}.
+ */
+public final class DefaultHostMetrics implements HostMetrics {
```
Comment (markelliot) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/DefaultHostMetrics.java]: package private?

**Review by markelliot — APPROVED** [review]

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/DefaultHostMetrics.java (hunk 7 above)
Comment (markelliot) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/DefaultHostMetrics.java]: (nvm, was looking at an old commit)

**Pull request merged**